src/main/java/com/alphawallet/ethereum/ERC721Token.java

```diff
@@ -0,0 +1,53 @@
+package com.alphawallet.ethereum;
+
+import com.alphawallet.attestation.Attestation;
+import com.alphawallet.attestation.core.ASNEncodable;
+import com.alphawallet.token.tools.Numeric;
+import org.bouncycastle.asn1.*;
+import org.bouncycastle.asn1.x509.AlgorithmIdentifier;
+
+import java.io.IOException;
+import java.math.BigInteger;
+
+public class ERC721Token implements ASNEncodable {
+    public String address;
+    public BigInteger tokenId;
+
+    public ERC721Token(String address, String tokenId)
+    {
+        this.address = address;
+        try
+        {
+            this.tokenId = new BigInteger(tokenId);
+        }
+        catch (Exception e)
+        {
+            this.tokenId = BigInteger.ZERO;
+        }
+    }
+
+    public ERC721Token(String address, BigInteger tokenId)
+    {
+        this.address = address;
+        this.tokenId = tokenId;
+    }
+
+    public ASN1EncodableVector getTokenVector()
+    {
+        ASN1EncodableVector data = new ASN1EncodableVector();
+        data.add(new DEROctetString(Numeric.hexStringToByteArray(address)));
+        data.add(new ASN1Integer(tokenId));
```

Comment:
Why is this encoded as an integer and not just a 256 bit string? (It can of course be both but I am asking in relation to the asn 1 definitions, whether we should do one or the other :) )

Reply:
Because it's much faster to decode in Solidity.

If it's encoded as a string, we need to convert it back from ASCII to hex then to address. I have written this code in assembler as the gas cost was quite high for compiled Solidity. Previously we encoded the address as an ASCII string - so a pair of ASCII character becomes a nibble; two pairs a byte.

If it's an integer then it's a fraction of the cost, and can be done by pulling out the bytes directly, then manipulating the bytes to fit into an address variable, which is very cheap and clean.

Reply:
Okay, but in that case shouldn't the address also be an integer? 
Maybe we are misunderstanding each other, I don't mean an ASCII string. I mean an OctetString, just like you encode the  address here. The reason I am asking is because this allows us to specify a specific amount of bytes in ASN1, whereas an ASN1 Integer is unbounded.

Reply:
> Okay, but in that case shouldn't the address also be an integer?
> Maybe we are misunderstanding each other, I don't mean an ASCII string. I mean an OctetString, just like you encode the address here. The reason I am asking is because this allows us to specify a specific amount of bytes in ASN1, whereas an ASN1 Integer is unbounded.

Ah I see. OK let's change it to OctetString. Address could be an integer and that would be ok too - as bytes it's equally as easy to convert to an address.